### PR TITLE
Instrument significant timings, rule counts, and error counts

### PR DESCRIPTION
**Pull request by ndonze-fs**
This PR uses telemetry to instrument the following:
- Overall DLO initialization time
- Time to collect configured rules from the `window` object
- Time to register all rules
- Time to handle each event (execute operator pipeline and send to configured rule destination) along with the number of operators and type of operators configured for a given event
- Count of client errors segmented by error type
- Count of rules configured for the DLO instance

### CODE REVIEW

File: rollup.config.js
Comment (ndonze-fs) [inline rollup.config.js:30]: Changing the module resolution strategy was necessary to import `package.json` via the `@rollup/plugin-json` plugin. Versioning DLO is already a little burdensome:
- `npm version`
- Add `CHANGELOG` entry
- Update version in `README`
- After PR merges, tag merged version and push new version tag

I wanted to lean on the version from `package.json` to avoid adding another step (e.g. update version in `telemetry.ts`).

File: src/embed/init.impl.ts
Comment (ndonze-fs) [inline src/embed/init.impl.ts:92]: I'm interested in feedback on the overall pattern here. We could take at least a couple alternative directions:
- Add a `Telemetry.error` or similar method which handles adding the error type attribute. Would allow calling e.g. `Telemetry.error(errorType.ruleRegistrationError)`
- Report each error as a separate count and avoid the attribute altogether. We'll probably end up aggregating these, but that can be done in our custom exporter. Example if we reported each error as a separate count: `Telemetry.count(telemetryType.ruleRegistrationError, 1)`

File: src/handler.ts
Comment (ndonze-fs) [inline src/handler.ts:64]: Is this worth collecting? Figured knowing which operators are used and to what degree would be valuable. Is comma separated the best way to go? Should these at least be alphabetically sorted for consistency?

File: src/observer.ts
Comment (ndonze-fs) [inline src/observer.ts:141]: Per the comment, removing this logging statement causes test failures. I suspect there's some global test state that isn't being handled properly. Rather than resolving in this PR which already has a decent sized scope, I plan to address this when we deprecate the `record` log level.

File: src/utils/telemetry.ts
Comment (ndonze-fs) [inline src/utils/telemetry.ts:21]: I elected to have separate error constants for telemetry rather than overloading e.g. `LogMessageType.OperatorError` so these could evolve independently. Logging message types could be changed with less friction; telemetry error types should really be pinned down as exporters may implement logic based on those values.

In the current implementation, I'm also explicitly calling `Telemetry.count` after we call `Logger.getInstance().error`. We could consider always calling `Telemetry.count` within `Logger.getInstance().error`, at which point I'd probably lean toward overloading `LogMessageType` or maybe even extracting that to a dedicated `ErrorType` concept which is used by both logging and telemetry. Interested in feedback on this.

File: src/embed/init.impl.ts
Comment (patrick-fs) [reply src/embed/init.impl.ts:92]: I like the `Telemetry.error` interface - usage is more explicit.

I have some questions about using the error type as the metric name: any concerns about Cardinality issues? Would using the error type this way help when reporting things in grafana (e.g. if we wanted to have a graph of all errors, visualized by type over time)?

File: src/embed/init.impl.ts
Comment (patrick-fs) [inline src/embed/init.impl.ts:113]: Would it be a good idea to add a comment here about what the expected `Telemetry` metric collection behavior is if no `win._dlo_telemetryExporter` has been defined?

File: src/embed/init.impl.ts
Comment (patrick-fs) [reply src/embed/init.impl.ts:72]: I think it would be interesting to keep it in and look for any unexpected outliers in production initializations.

File: src/handler.ts
Comment (patrick-fs) [reply src/handler.ts:64]: I think insights about operators is interesting and including the operator chain in the event-handling performance metric should give us insights into whether there are any unexpected latencies in our rule sets - could be interesting for tuning default rules or understanding performance implications of custom rules.

File: src/handler.ts
Comment (patrick-fs) [inline src/handler.ts:67]: Per my comment above, maintaining the order in which operators are executed (and all operators in the execution chain) could be insightful. Does this report all operators across all rules, rather than just the ones that apply to this particular event?

File: src/utils/telemetry.ts
Comment (patrick-fs) [reply src/utils/telemetry.ts:21]: I think we'll need to revisit how our current logger works at some point - currently we're consuming customer FS.event quota. I can see a future where both logging and metrics are collapsed into a common framework that uses the same telemetry provider - maybe one that implements both the current otel-based interfaces as well as a "logger" interface. Idunno, just riffing; I could be way off-base about implementation.

I think the implementation as-is is fine for this PR.

File: src/utils/telemetry.ts
Comment (patrick-fs) [reply src/utils/telemetry.ts:354]: I suggest sticking with SemVar - if it's a breaking change (even if no one is likely using it) then do a Major version bump.

File: src/utils/telemetry.ts
Comment (patrick-fs) [inline src/utils/telemetry.ts:261]: What do you think about naming this `mergeDefaultAttributes`?

File: src/utils/telemetry.ts
Comment (patrick-fs) [inline src/utils/telemetry.ts:265]: I wonder if you could do something like:

```
const mergedAttributes = this.defaultAttributes
for (const [key, value] of Object.entries(attributes)) {
  mergedAttributes[key] = value;
}
return mergedAttribute;
```

File: src/embed/init.impl.ts
Comment (ndonze-fs) [reply src/embed/init.impl.ts:92]: This is a good question. We have some precedent for tracking error type as a label with errors in Grafana, and doing so would allow us to pivot by error type in Grafana. Even if we capture error type as metric metadata here, we can also make an independent decision about whether to include error type as a label in Grafana. For example, we could drop/ignore error type or choose to log to another sink in our exporter and downstream processing.

Comment (patrick-fs) [reply src/embed/init.impl.ts:92]: Yep - good point.

File: src/handler.ts
Comment (ndonze-fs) [reply src/handler.ts:67]: Only rules applying to the event being processed (from the defined rule which triggered the event) would be included. I could see the argument for not de-duping like you mentioned and preserving original processing order.

File: src/utils/telemetry.ts
Comment (ndonze-fs) [reply src/utils/telemetry.ts:265]: This or object literal spreading would be a lot simpler. I've been trying to avoid APIs not supported by IE11 since we still have some open work to remove a few lingering IE11 errors in DLO. If we decide to no longer support IE11 in DLO or add something like babel to build process we could use `Object.assign` or object literal spreading.

File: src/handler.ts
Comment (patrick-fs) [reply src/handler.ts:67]: Ah, got it - I forgot that there's one `DataHandler` instance per rule. Yeah I think not de-duping would be better. The data is interesting and I don't think de-duping would save much in terms of payload size.

File: src/utils/telemetry.ts
Comment (patrick-fs) [reply src/utils/telemetry.ts:265]: I see, I just straight missed this comment:

```
// We're avoiding Object.assign and object literal spreading since they're
// not supported by IE11
```

Comment (ndonze-fs) [reply src/utils/telemetry.ts:265]: I'm glad you asked about this. I learned we can use object literal spreading and TypeScript (based on our compiler options) will use and polyfill `Object.assign`. We can simplify this function to:

```
  private insertDefaultAttributes(attributes?: Attributes): Attributes {
    return { ...this.defaultAttributes, ...attributes };
  }
```

When built, this results in:

```
    DefaultTelemetryProvider.prototype.insertDefaultAttributes = function (attributes) {
        return __assign(__assign({}, this.defaultAttributes), attributes);
    };

    ...

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
```

Minified build size is roughly the same, though full build size is ~1KB smaller.

**Comment by ndonze-fs** [comment]
@patrick-fs This should be ready for another review. I bumped the major version and added a corresponding entry to the `CHANGELOG`. I haven't updated the `README` to point to the latest version yet. I'd prefer to wait until we release because we'll probably need to shift both the version specific and latest links to `/v2`, and these links won't be valid until we release.

File: src/embed/init.impl.ts
Comment (patrick-fs) [reply src/embed/init.impl.ts:92]: Just pinging here in case you intended to convert this to a `Telemetry.error` call. Up to you.

**Review by patrick-fs — APPROVED** [review]
LGTM! I left one remark about `Telemetry.error` in an existing comment thread. Feel free to merge as-is, if you'd like.

**Pull request merged**
